File: Python/12. 서버저장조건 실시간검색.py
import asyncio
import ebest
from common import *
from app_keys import *

async def main():
    api=ebest.OpenApi()
    api.on_message = on_message
    api.on_realtime = on_realtime
    if not await api.login(appkey, appsecretkey): return print(f'연결실패: {api.last_message}')
    if api.is_simulation: return print('실시간검색은 실서버에서만 가능합니다')
     
    # 조건검색식 리스트 조회
    request = {
        't1866InBlock': {
            'user_id': user_id, # 사용자ID 8자리
            'gb': '0', # 0 : 그룹+조건리스트 조회, 1 : 그룹리스트조회, 2 : 그룹명에 속한 조건리스트조회
            'group_name': '',
            'cont': '0',
            'cont_key': '',
        }
    }
    response = await api.request('t1866', request)
    if not response: return print(f'요청실패: {api.last_message}')
    
    cond_list = response.body['t1866OutBlock1']
    print_table(cond_list)
    
    # 요청할 조건검색식 선택
    cond_len = len(cond_list)
    sel_index = int(input(f'조건검색식index (0~{cond_len-1})를 입력하세요:'))
    if sel_index >= cond_len: return print('잘못된 index')
    
    # 조건검색식 조회
    query_index = cond_list[sel_index]['query_index']
    request = {
        't1859InBlock': {
            'query_index': query_index,
        }
    }
    response = await api.request('t1859', request)
    if not response: return print(f'요청실패: {api.last_message}')
    
    item_list = response.body.get('t1859OutBlock1', None)
    if item_list:
        print_table(item_list)
    else:
        print(f'조건검색식[{query_index}] 결과 없음')
    
    # 실시간검색 등록
    request = {
        't1860InBlock': {
            'sSysUserFlag': 'U', # 'U' 고정
            'sFlag': 'E', # 'E:'등록, 'D':중지
            'sAlertNum': '', # Flag 값 'D':중지 일떄만 입력 - 등록 요청 시 수신받은 t1860OutBlock.sAlertNum 값
            'query_index': query_index,
        }
    }
    response = await api.request('t1860', request)
    if not response: return print(f'요청실패: {api.last_message}')
    
    real_req_data = response.body['t1860OutBlock']
    print_table(real_req_data)
    
    sAlertNum:str = real_req_data['sAlertNum']
    if sAlertNum == '':
        print('실시간검색 등록실패')
    else:
        print('실시간검색 등록성공, 5분동안 작동...')
        await api.add_realtime('AFR', sAlertNum)
        await asyncio.sleep(5*60) # 5분 동안 유효, 후에 중지
        # 실시간검색 중지
        await api.remove_realtime('AFR', sAlertNum)
        request = {
            't1860InBlock': {
                'sSysUserFlag': 'U', # 'U' 고정
                'sFlag': 'D', # 'E:'등록, 'D':중지
                'sAlertNum': sAlertNum, # Flag 값 'D':중지 일떄만 입력 - 등록 요청 시 수신받은 t1860OutBlock.sAlertNum 값
                'query_index': query_index,
            }
        }
        response = await api.request('t1860', request)
        await asyncio.sleep(1)
    
    ... # 다른 작업 수행
    await api.close()

def on_message(api:ebest.OpenApi, msg:str): print(f'on_message: {msg}')

def on_realtime(api:ebest.OpenApi, trcode, key, realtimedata):
    if trcode == 'AFR':
        print(f'실시간조건검색: {trcode}, {key}, {realtimedata}')

asyncio.run(main())

File: CSharp/12. 서버저장조건 실시간검색.cs
using System.Text.Json;

namespace CSharp;

internal class _12 : SampleBase
{
    public override async Task ActionImplement()
    {
        // [요청] t1866 : 서버저장조건리스트조회(API)
        t1866 tr_data = new()
        {
            t1866InBlock = new(Secret.UserId, "0", "", "0", ""),
        };
        await api.GetTRData(tr_data);
        if (tr_data.t1866OutBlock is null)
        {
            // 오류 처리
            print(tr_data.rsp_cd.Length > 0 ? $"{tr_data.rsp_cd}-{tr_data.rsp_msg}" : api.LastErrorMessage);
            return;
        }

        // tr_data.t1866OutBlock 데이터 처리
        print(tr_data.t1866OutBlock);
        print(tr_data.t1866OutBlock1);

        if (tr_data.t1866OutBlock1 == null)
        {
            print("서버에 저장된 조건검색식이 없습니다.");
            return;
        }

        // 요청할 조건검색식 선택
        int sel_index;
        do
        {
            Console.Write($"조건검색식index (0~{tr_data.t1866OutBlock.result_count - 1})를 입력하세요:");
            int.TryParse(Console.ReadLine(), out sel_index);
            if (sel_index < 0 || sel_index >= tr_data.t1866OutBlock.result_count)
            {
                print("잘못된 입력입니다.");
                sel_index = -1;
            }
        } while (sel_index == -1);

        // [요청] t1859 : 서버저장조건 조건검색
        var query_index = tr_data.t1866OutBlock1[sel_index].query_index;
        t1859 tr_data_cond = new()
        {
            t1859InBlock = new(tr_data.t1866OutBlock1[sel_index].query_index),
        };
        await api.GetTRData(tr_data_cond);
        if (tr_data_cond.t1859OutBlock is null)
        {
            // 오류 처리
            print(tr_data_cond.rsp_cd.Length > 0 ? $"{tr_data_cond.rsp_cd}-{tr_data_cond.rsp_msg}" : api.LastErrorMessage);
            return;
        }

        // tr_data_cond.t1859OutBlock 데이터 처리
        print(tr_data_cond.t1859OutBlock);
        print(tr_data_cond.t1859OutBlock1);

        // [요청] t1860 : 서버저장조건 실시간검색
        t1860 tr_data_real = new()
        {
            t1860InBlock = new("U", "E", "", query_index),
        };
        await api.GetTRData(tr_data_real);
        if (tr_data_real.t1860OutBlock is null)
        {
            // 오류 처리
            print(tr_data_real.rsp_cd.Length > 0 ? $"{tr_data_real.rsp_cd}-{tr_data_real.rsp_msg}" : api.LastErrorMessage);
            return;
        }

        // tr_data_real.t1860OutBlock 데이터 처리
        print(tr_data_real.t1860OutBlock);

        // 실시간 시세등록
        api.OnRealtimeEvent += Api_OnRealtimeEvent;
        if (!await api.AddRealtimeRequest("AFR", tr_data_real.t1860OutBlock.sAlertNum))
        {
            print($"AFR 실시간 등록 오류:{api.LastErrorMessage}");
            return;
        }

        // 10분후 리턴
        print("10분동안 실시간 조건검색 작동중...");
        await Task.Delay(600000);
    }

    private static void Api_OnRealtimeEvent(object? sender, eBEST.OpenApi.Events.EBestOnRealtimeEventArgs e)
    {
        if (e.TrCode.Equals("AFR")) // API사용자조건검색실시간
        {
            var real_cond = e.RealtimeBody.Deserialize<AFROutBlock>(); // JsonSerializer.Deserialize<AFROutBlock>(e.RealtimeBody);
            print($"{e.TrCode}, {e.Key}");
            print(real_cond);
        }
    }
}
